### PR TITLE
[controller] Check dvc heartbeat in all regions & update version swap logic for deferred swap service

### PR DESCRIPTION
**Pull request by misyel**
## Problem Statement

**1. Davinci heartbeat check in deferred version swap service + parent admin**
Currently, in the deferred version swap service, we check for a dvc heartbeat for only one target region before skipping the store, but a dvc client can only be present in some regions and not all regions. To accurately skip a dvc store, we will need to check for a dvc heartbeat in *all* regions instead.

**2. Version swap logic in deferred version swap service**
Currently, the version swaps happen in the following cases:

1. Version status is `PUSHED` (this is set upon successful push in *all* regions)
2. Target region is complete in > 50% of the target regions
3. Non target regions have all reached terminal status
4. Non target regions is complete in > 50% of non target regions

And it fails to account for cases where:

1. Push is successful in *some* regions, but not all (version status is marked as `KILLED`)

**3. Code readability in deferred version swap service**
The code in the deferred version swap service has gotten to a point where there are a lot of checks and early exits and it is difficult to understand what is happening in the code at a glance

## Solution

**Overview of changes (green represents changed logic):**
Blank diagram

**1. Davinci heartbeat check in deferred version swap service + parent admin**
Query all regions and check for a dvc heartbeat instead of only checking *one* target region. If there is a dvc heartbeat reported in *any* region, we will skip the store and not perform a version swap

**2. Version swap logic in deferred version swap service**
Update the version swap logic to check for KILLED parent version statuses to account for push completions in only *some* regions. This changes the version swap logic to perform version swaps in cases like this:

<table>
<tr><td>

|regions| push status | final version|
|--|--|--|
|region1(target region)| completed| 1 |
|region2| failed|0 |
|region3| completed|1|

</td><td>

|regions| push status | final version|
|--|--|--|
|region1(target region)| completed| 1 |
|region2(target region)| failed|0 |
|region3| completed|1|

</td></tr> </table>

<table>
<tr><td>

|regions| push status | final version|
|--|--|--|
|region1(target region)| completed| 1 |
|region2| failed|0 |
|region3| failed|0|

</td><td>

|regions| push status | final version|
|--|--|--|
|region1(target region)| failed| 0 |
|region2(target region)| failed|0 |
|region3| completed|0|

</td></tr> </table>

**3. Code readability in deferred version swap service**

1. Refactored logical chunks of code into separate methods to improve readability & added block comments to each method
2. Added more thorough e2e testing to account for edge cases (failing pushes in target regions, multiple target regions, etc)

###  Code changes
- [ ] Added new code behind **a config**. If so list the config names and their default values in the PR description.
- [x] Introduced new **log lines**. 
  - [x] Confirmed if logs need to be **rate limited** to avoid excessive logging.

###  **Concurrency-Specific Checks**
Both reviewer and PR author to verify
- [x] Code has **no race conditions** or **thread safety issues**.
- [x] Proper **synchronization mechanisms** (e.g., `synchronized`, `RWLock`) are used where needed.
- [x] No **blocking calls** inside critical sections that could lead to deadlocks or performance degradation.
- [x] Verified **thread-safe collections** are used (e.g., `ConcurrentHashMap`, `CopyOnWriteArrayList`).
- [x] Validated proper exception handling in multi-threaded code to avoid silent thread termination.

## How was this PR tested?

- [ ] New unit tests added.
- [x] New integration tests added.
- [x] Modified or extended existing tests.
- [ ] Verified backward compatibility (if applicable).

## Does this PR introduce any user-facing or breaking changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Clearly explain the behavior change and its impact.

### CODE REVIEW

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 1]
```diff
@@ -82,40 +86,254 @@ private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegio
...
+        return true;
+      }
+
+      StoreInfo targetRegionStore = targetRegionStoreResponse.getStore();
+      Optional<Version> version = targetRegionStore.getVersion(targetVersionNum);
+      if (!version.isPresent()) {
+        String message =
+            "Unable to find version " + targetVersionNum + " for store: " + storeName + " in region " + region;
+        logMessageIfNotRedundant(message);
+        return false;
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:155]: why return false here? Also if we cannot find target version, may be we can check the previous version for the dvc flag and return true in that case.

Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java]: Good catch, we shouldn't return false early here yet as other regions may have a dvc heartbeat. Updated logic to check for the previous version too if the target version does not exist and removed the return statement

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 2]
```diff
@@ -82,38 +86,263 @@ private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegio
...
+        if (version.get().getIsDavinciHeartbeatReported()) {
+          String message = "Skipping version swap for store: " + storeName + " on version: " + targetVersionNum
+              + " as there is a davinci heartbeat in region: " + region;
+          logMessageIfNotRedundant(message);
+          return true;
+        }
+      }
+
+      // Check the previous version for a dvc heartbeat if we can't find the target version number
+      if (!version.isPresent() && previousVersion.isPresent()) {
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:167]: nit: check `!version.isPresent()` is not needed here right?

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java [hunk 3]
```diff
@@ -3774,6 +3789,7 @@ private OfflinePushStatusInfo getOffLineJobStatus(
           }
           // status PUSHED is set when batch store's target region push is completed, but other region are yet to
           // complete
+          // somehow removing the outer if makes the status ERROR instead of KILLED
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java:3792]: what this comment for?

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 4]
```diff
@@ -82,38 +86,263 @@ private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegio
...
+    return true;
+  }
+
+  /**
+   * Given a list of regions and push statuses, return a list of regions whose push statuses are COMPLETED or ERROR
+   * @param regions list of regions to find the push status for
+   * @param pushStatusInfo wrapper containing push status info
+   * @return
+   */
+  private Map<String, Set<String>> getCompletedAndFailedRegions(
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:254]: Can we just return some count instead? We are allocating too any temporary objects couple of times in a loop for each store. Only time you need the actual set instead of size is in `completedNonTargetRegions` which we can just iterate over.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java (hunk 2 above)
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:166]: It is needed to account for cases where both the previous version and target version is present, but there is no dvc heartbeat detected in the target version and there is a dvc heartbeat detected in the previous version. In this case, we should still return false as the current version dvc heartbeat status is more recent than the previous version dvc heartbeat status. We should only use the previous version's dvc heartbeat state if we can't find the target version

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java (hunk 3 above)
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java]: removed

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java (hunk 4 above)
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java]: Good point - updated the function to return an int representing the number of regions w/ that specific status instead to avoid unnecessary temporary objects

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 5]
```diff
@@ -82,38 +85,265 @@ private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegio
...
+    if (targetVersion == null) {
+      return false;
+    }
+
+    String targetRegionsString = targetVersion.getTargetSwapRegion();
+    if (StringUtils.isEmpty(targetRegionsString)) {
+      return false;
+    }
+
+    // The store is eligible for a version swap if its push job is in terminal status. For a target region
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:237]: Can you rewrite the javadoc to explain why we need to rollforward even if the push is KILLED?

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 6]
```diff
@@ -82,38 +85,265 @@ private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegio
...
+    // push, the parent version status is set to PUSHED in getOfflinePushStatus when this happens or KILLED if the push
+    // failed
+    if (targetVersion.getStatus() != VersionStatus.PUSHED && targetVersion.getStatus() != VersionStatus.KILLED) {
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:240]: We dont want to reprocess a previously killed pushjob store right?

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 7]
```diff
@@ -130,173 +360,73 @@ public void run() {
             for (Store store: stores) {
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:360]: nit: can you rename to parentStore ?

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java (hunk 5 above)
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:237]: It will not be a continuous loop. If all regions fail, we will mark the parent version status as `ERROR` so it will not meet the conditions above again to enter the loop for that version. See the the change [here](https://github.com/linkedin/venice/blob/44d051c4227b96d5764345ea8b9e9ef0995c6617/services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java#L289)

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java (hunk 6 above)
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:240]: Yeah, we won't reprocess a previously killed push job if we already processed it once and have come to the conclusion that:
1. All target regions failed. In this case we will mark the status as `ERROR` so it doesn't enter the loop again
2. more than 50% target regions have failed. In this case we will mark the status as `PARTIALLY_ONLINE`
3. We perform a roll forward on 1+ non target regions. In this case, we will mark the status as `PARTIALLY_ONLINE`

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 8]
```diff
@@ -82,38 +88,268 @@ private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegio
...
+    int numFailedTargetRegions = getRegionsWithPushStatusCount(targetRegions, pushStatusInfo, ExecutionStatus.ERROR);
+    if (numFailedTargetRegions > targetRegions.size() / 2) {
+      String message = "Skipping version swap for store: " + store.getName() + " on version: " + targetVersionNum
+          + " as push failed in the majority of target regions. Completed target regions: " + numCompletedTargetRegions
+          + " , failed target regions: " + numFailedTargetRegions + ", target regions: " + targetRegions;
+      logMessageIfNotRedundant(message);
+      store.updateVersionStatus(targetVersionNum, ERROR);
+      repository.updateStore(store);
+      return true;
+    } else if (numFailedTargetRegions + numCompletedTargetRegions != targetRegions.size()) {
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:298]: we can skip this else part

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java [hunk 9]
```diff
@@ -126,177 +362,82 @@ public void run() {
...
-                LOGGER.warn(
-                    "Skipping version swap for store: {} on version: {} as majority of non target regions have failed",
-                    storeName,
-                    targetVersionNum);
-                store.updateVersionStatus(targetVersionNum, ERROR);
-                repository.updateStore(store);
+              Set<String> remainingRegions = getRegionsForVersionSwap(coloToVersions, targetRegions);
+
+              // Check if push is successful in majority target regions
+              if (didPushFailInTargetRegions(
```
Comment (majisourav99) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java:408]: We enter here only if push status is KILLED

File: internal/venice-common/src/main/java/com/linkedin/venice/meta/VersionStatus.java [hunk 10]
```diff
@@ -8,21 +8,51 @@
...
+  NOT_CREATED(0),
+
+  /**
+   * Version has been pushed to venice(offline job related to this version has been completed), but is not ready to
+   * serve read request because the writes to this store is disabled
+   */
+  STARTED(1),
+
+  /**
+   * Version has been pushed to venice and is ready to serve read request. Intermediate status after a push job succeeds before DeferredVersionSwapService
```
Comment (majisourav99) [inline internal/venice-common/src/main/java/com/linkedin/venice/meta/VersionStatus.java:26]: Intermediate status after a push job finished pushing data to the child colo before DeferredVersionSwapService flips the status to ONLINE. This status only exists in the parent

File: internal/venice-common/src/main/java/com/linkedin/venice/meta/VersionStatus.java [hunk 11]
```diff
@@ -8,21 +8,51 @@
...
-  NOT_CREATED(0), STARTED(1),
-  // Version has been pushed to venice(offline job related to this version has been completed), but is not ready to
-  // serve read request because the writes to this store is disabled.
+  /**
+   * This version hasn't been created yet. It is not persisted in ZK
+   */
+  NOT_CREATED(0),
+
+  /**
+   * Version has been pushed to venice(offline job related to this version has been completed), but is not ready to
```
Comment (majisourav99) [inline internal/venice-common/src/main/java/com/linkedin/venice/meta/VersionStatus.java:20]: Version has been created and started ingesting new data, but has not completed ingestion and not read to serve read traffic

**Review by majisourav99 — APPROVED** [review]
LGTM! Thanks a lot @misyel
